liboptparse: Support setting alternate usage callback function

chu11 opened the pull request:
I thought it was sort of inconvenient to remove the default "help" option and then re-add it back just so we can set an alternate (or no) usage callback function.  Since I was duplicating the code in ```cmd/flux``` into ```cmd/flux-kvs```, figured could refactor a bit.  So I added an option to allow the user to set an alternate usage callback function.  The callback can be disabled by setting the callback to NULL.

Also fix one nit along the way.

grondo commented:
I'm a bit worried that this breaks the abstraction that `--help` is just another option, as you've elevated it to something special. At least, an error should be returned if the "help" option isn't found in `optparse_set_usage_fn`. An optparse object is not guaranteed to have a `--help` option.

Also, since this set item is setting the callback for the `--help` option explicitly, maybe the item should be called OPTPARSE_HELP_FN? Should the `--help` option and handler be special-cased in other ways as well?

chu11 commented:
Ahh, I see about the abstraction aspect now.  Thinking about this a tad.  Perhaps instead some common/convenience liboptparse functions need to be developed?  Minimally, we have two (three w/ future ```flux-kvs``` fixups) "help opt" declarations sitting around.

```
struct optparse_option helpopt = {
      .name = "help", .key = 'h', .usage = "Display this message"
};
```

It seems like unnecessary duplication.

chu11 commented:
Thinking a tad more, perhaps I jumped the gun on this enhancement.  With more tools using ```liboptparse``` in the future, we can figure out some of the common-ness later on.

grondo commented:
What if instead of special case for `--help`, you made the new `optparse_set` item take the option name and a callback function, e.g.

```
    optparse_err_t e = optparse_set (p, OPTPARSE_OPTION_CB, "help", NULL);
```

or just a special function

```
   optparse_cb_f oldcb = optparse_set_option_cb (p, "help", NULL);
```

Maybe it would be useful to return a copy of the "old" callback, I'm not sure. This removes the special case for `--help` option, so I'd be happier with it.

Of course, we could also have a function that returns the `struct optparse_option *` for a given string, then it could be manipulated by the caller at will, but that seems a little bit dangerous.

chu11 commented:
I like this one:

```
    optparse_err_t e = optparse_set (p, OPTPARSE_OPTION_CB, "help", NULL);
```

I'll try and tweak to do this one instead.

chu11 commented:
just re-pushed with changes.  Added error handling so ```OPTPARSE_BAD_ARG``` is returned when appropriate and added some extra tests for checking for bad input.

grondo commented on src/cmd/flux.c:
```diff
@@ -97,13 +94,8 @@ static optparse_t * setup_optparse_parse_args (int argc, char *argv[])
-        log_msg_exit ("optparse_add_option (\"help\")");
+    // Disable automatic `--help' in favor of our own usage() from above
+    optparse_set (p, OPTPARSE_OPTION_CB, "help", NULL);
 
```
Though it is almost impossible to have an error, should we check error here just for completeness? (And in case code  gets duplicated or used as an example)

chu11 replied:
Yup, you're right.  I was being lazy.  New push coming.

grondo commented:
Just one minor comment, otherwise good improvement, thanks!

grondo commented:
LGTM! Nice improvement over the old code, thanks.
